### PR TITLE
bpo-35279: reduce default max_workers of ThreadPoolExecutor

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue35279

### CODE REVIEW

**Review by asvetlov — APPROVED** [review]
LGTM!

File: Doc/library/concurrent.futures.rst [hunk 1]
```diff
@@ -159,6 +159,15 @@ And::
    .. versionchanged:: 3.7
       Added the *initializer* and *initargs* arguments.
 
+   .. versionchanged:: 3.8
+      Default value of *max_workers* is changed to ``min(32, os.cpu_count() + 4)``.
+      This default value preserves at least 5 worker for I/O bound task.  It utilize
```
Comment (tirkarthi) [inline Doc/library/concurrent.futures.rst]: Not a native English speaker but I think it should be "It utilizes" and "And it avoids" in the next sentence.

Comment (csabella) [reply Doc/library/concurrent.futures.rst]: Also, `at least 5 workers`.  And I'm not sure if it should be `for I/O bound tasks` or 'for the I/O bound task`.

File: Doc/library/concurrent.futures.rst [hunk 2]
```diff
@@ -159,6 +159,15 @@ And::
...
+      at most 32 CPU cores for CPU bound task which release GIL.  And it avoid using
```
Comment (csabella) [inline Doc/library/concurrent.futures.rst]: `for CPU bound tasks` and `which release the GIL`.

File: Doc/library/concurrent.futures.rst [hunk 3]
```diff
@@ -159,6 +159,15 @@ And::
...
+      very large resource implicitly on many core machines.
```
Comment (csabella) [inline Doc/library/concurrent.futures.rst]: s/resource/resources

File: Doc/library/concurrent.futures.rst [hunk 4]
```diff
@@ -159,6 +159,15 @@ And::
    .. versionchanged:: 3.7
       Added the *initializer* and *initargs* arguments.
 
+   .. versionchanged:: 3.8
+      Default value of *max_workers* is changed to ``min(32, os.cpu_count() + 4)``.
+      This default value preserves at least 5 workers for I/O bound tasks.
+      It utilizes at most 32 CPU cores for CPU bound tasks which release the GIL.
+      And it avoids using very large resources implicitly on many-core machines.
+
+      ThreadPoolExecutor now reuses idle worker threads before starting
```
Comment (wyaadarsh) [inline Doc/library/concurrent.futures.rst:168]: Where is the corresponding change in the code? For reusing the idle worker threads before starting *max_workers* worker threads?

Comment (methane) [reply Doc/library/concurrent.futures.rst:168]: https://github.com/python/cpython/pull/6375

**Pull request merged**